### PR TITLE
#4790 Add onboarding Marketplace Blueprint info to Get Started page

### LINKED ISSUE #4790: Modify content on extension console "getting started"
As a means of improving the onboarding workflow coming from activating a blueprint, we need to look at the content on the extension console "getting started" tab and provide appropriate instructions to direct the user based on their situation\. Corinne will help Misha with specific content changes\.

<p></p>

<https://www.notion.so/pixiebrix/Onboard-via-Marketplace-Blueprint-df00e33f9b114676b5440191a447ccee>

### PR DESCRIPTION
**Pull request by mnholtz**
## What does this PR do?

- Closes #4790 
- Depends on back-end change to Milestones https://github.com/pixiebrix/pixiebrix-app/pull/2803

## Reviewer Tips

- Main functional changes are tweaks to the useOnboardingTabs hook in `BlueprintsPageSidebar`, the addition of copy to `GetStartedView`, and the addition of the milestone creation for `first_time_public_blueprint_install` in `useInstall`
- On the back-end, we are adding a `value` field to `Milestone`s

## Discussion

- I've been considering replacing the `hasSomeBlueprintEngagement` logic in `useOnboardingTabs` with milestones for a while. This requires a little more thought & time however

## Demo

https://user-images.githubusercontent.com/36575242/209226831-c09d9845-8522-4c53-abc5-977ac65f4564.mov

## Checklist

- [x] Add tests
- [x] Run Storybook and manually confirm that all stories are working
- [x] Designate a primary reviewer @BALEHOK or @BLoe

### CODE REVIEW

File: src/hooks/useMilestones.ts [hunk 1]
```diff
@@ -18,27 +18,44 @@
...
 
 function useMilestones(): MilestoneHelpers {
-  const milestones = useSelector(selectMilestones);
+  const cachedMilestones = useSelector(selectMilestones);
+  const { data: me, isFetching, isLoading, refetch } = useGetMeQuery();
+  const milestones = me ? me.milestones : cachedMilestones;
 
   return useMemo(() => {
-    const userMilestoneKeys = new Set((milestones ?? []).map((x) => x.key));
+    const milestonesByKey = Object.fromEntries(
```
Comment (twschiller) [inline src/hooks/useMilestones.ts:39]: Should prefer `Map` to using object for lookups

File: src/hooks/useMilestones.ts [hunk 2]
```diff
@@ -18,27 +18,44 @@
...
 
     return {
+      getMilestone,
       hasMilestone,
       hasEveryMilestone,
+      isFetching,
+      isLoading,
+      refetch,
     };
   }, [milestones]);
```
Comment (twschiller) [inline src/hooks/useMilestones.ts:60]: I don't think this dependency list is right? The isFetching won't change if refetch is called

File: src/hooks/useMilestones.test.tsx [hunk 3]
```diff
@@ -0,0 +1,128 @@
...
+    preloadedState: initialState,
+  });
+}
+
+jest.mock("@/services/api", () => ({
+  useGetMeQuery: jest.fn(),
+}));
+
+const renderUseMilestones = (initialState?: any) => {
+  (useGetMeQuery as jest.Mock).mockImplementation(() => ({
```
Comment (twschiller) [inline src/hooks/useMilestones.test.tsx:41]: Shouldn't this be returning `{data: {milestones}, isFetching, ...}` to match the return value of useGetMeQuery?

File: src/options/pages/blueprints/BlueprintsPageSidebar.tsx [hunk 4]
```diff
@@ -103,10 +106,15 @@ const useOnboardingTabs = (
     blueprintsSlice.actions.setActiveTab
   );
   const { data: installableViewItems } = tableInstance;
-  const { data: me, isLoading: isMeLoading } = useGetMeQuery();
+  const { data: me, isLoading: isMeLoading, isFetching } = useGetMeQuery();
```
Comment (twschiller) [inline src/options/pages/blueprints/BlueprintsPageSidebar.tsx:109]: Be consistent with naming for isMeLoading. Rename to isMeFetching

File: src/options/pages/blueprints/BlueprintsPageSidebar.tsx [hunk 5]
```diff
@@ -142,7 +154,7 @@ const useOnboardingTabs = (
   }, []);
 
   useEffect(() => {
-    if (isStarterBlueprintsLoading || isMeLoading) {
+    if (isStarterBlueprintsLoading || isMeLoading || isFetching) {
```
Comment (twschiller) [inline src/options/pages/blueprints/BlueprintsPageSidebar.tsx:157]: I think you need to add isMeFetching to the dependency list

File: src/options/pages/blueprints/BlueprintsToolbar.tsx [hunk 6]
```diff
@@ -97,90 +97,92 @@ const BlueprintsToolbar: React.FunctionComponent<{
...
-        />
-
-        {isSorted && (
-          <Button
-            variant="link"
-            size="sm"
-            onClick={() => {
-              const value = [{ id: sortBy[0].id, desc: !sortBy[0].desc }];
-              setSortBy(value);
+      {activeTab.hideToolbar ?? (
```
Comment (twschiller) [inline src/options/pages/blueprints/BlueprintsToolbar.tsx:100]: The null coalescing operator doesn't seem correct here. It should be using `!activateTab.hideToolbar &&` or using a ternary operator

File: src/options/pages/blueprints/GetStartedView.tsx [hunk 7]
```diff
@@ -47,105 +47,123 @@ const ExternalLink: React.VoidFunctionComponent<{
...
-        </p>
-      </Col>
-    </Row>
-  </div>
-);
+}> = ({ width, height }) => {
+  const { homepage_url: homepageUrl } = browser.runtime.getManifest();
+  const { getMilestone, isFetching } = useMilestones();
+
+  const onboardingBlueprintId = isFetching
```
Comment (twschiller) [inline src/options/pages/blueprints/GetStartedView.tsx:54]: Why does this need an isFetching guard? You don't ever want to use the milestone cache for some reason?

File: src/options/pages/blueprints/utils/useInstall.ts [hunk 8]
```diff
@@ -102,6 +109,22 @@ function useInstall(recipe: RecipeDefinition): InstallRecipe {
...
+            key: "first_time_public_blueprint_install",
+            metadata: {
+              blueprintId: recipe.metadata.id,
+            },
+          });
+
+          dispatch(setActiveTab(BLUEPRINTS_PAGE_TABS.getStarted));
+
+          // The cache invalidation caused by createMilestone doesn't
+          // trigger a refetch soon enough after redirecting back
```
Comment (twschiller) [inline src/options/pages/blueprints/utils/useInstall.ts:123]: Not sure I understand this comment. Could you clarify what "soon enough" means? The comment is saying that the milestones need to be fetched before the cache invalidation?

If there's a race, you likely need to await the `createMilestone` call?

File: src/services/api.ts [hunk 9]
```diff
@@ -403,7 +404,7 @@ export const appApi = createApi({
         { type: "StarterBlueprints", id: "LIST" },
       ],
     }),
-    createMilestone: builder.mutation<Milestone, { key: string }>({
+    createMilestone: builder.mutation<Milestone, Milestone>({
```
Comment (twschiller) [inline src/services/api.ts:407]: This doesn't seem quite right. It should be the milestone without the `user` field. Correct?

File: src/options/pages/blueprints/GetStartedView.tsx [hunk 10]
```diff
@@ -47,105 +47,123 @@ const ExternalLink: React.VoidFunctionComponent<{
 const GetStartedView: React.VoidFunctionComponent<{
```
Comment (twschiller) [inline src/options/pages/blueprints/GetStartedView.tsx:47]: Could we get the new into Storybook? Would be great to have these stories so we can understand the journey better an make any UI tweaks

File: src/hooks/useMilestones.ts (hunk 2 above)
Comment (mnholtz) [reply src/hooks/useMilestones.ts:60]: Oops, you're right. Did we remove lint warnings for dependency lists? 🤔 I had been leaning on those warnings.

File: src/services/api.ts (hunk 9 above)
Comment (mnholtz) [reply src/services/api.ts:407]: It's weird this wasn't throwing any errors before 😳

File: src/hooks/useMilestones.ts (hunk 2 above)
Comment (fregante) [reply src/hooks/useMilestones.ts]: The lint is still there, is this related to the IDE lint issues you were having?

Comment (mnholtz) [reply src/hooks/useMilestones.ts]: Probably 🙄 I haven't been prioritizing investigating that over my other commitments, but looks like I should soon.

File: src/options/pages/blueprints/GetStartedView.tsx (hunk 10 above)
Comment (mnholtz) [reply src/options/pages/blueprints/GetStartedView.tsx:47]: Added a story here for both views: https://github.com/pixiebrix/pixiebrix-extension/pull/4866/files/63bb6f1ae1e539c65aef4ab08ed8795497fcaf67..4173f0a2c0614eecae06154987bf6e08b26ea543, the default view and the additional Activate Blueprint copy.

File: src/options/pages/blueprints/BlueprintsPageSidebar.tsx [hunk 11]
```diff
@@ -103,10 +106,15 @@ const useOnboardingTabs = (
...
-  const { data: me, isLoading: isMeLoading } = useGetMeQuery();
+  const { data: me, isLoading: isMeLoading, isFetching } = useGetMeQuery();
   const { hasMilestone } = useMilestones();
   const { flagOn } = useFlags();
   const { isBotGamesBlueprintInstalled } = useInstallBotGamesBlueprint();
+  const { getMilestone } = useMilestones();
+
+  const onboardingBlueprintId = getMilestone(
+    "first_time_public_blueprint_install"
+  )?.metadata?.blueprintId as RegistryId;
```
[the remark this answers is not in the document]
Comment (mnholtz) [reply src/options/pages/blueprints/BlueprintsPageSidebar.tsx:121]: Technically metadata is optional, although also technically the server won't omit the property even if it is null in the database. https://github.com/pixiebrix/pixiebrix-extension/blob/4173f0a2c0614eecae06154987bf6e08b26ea543/src/types/swagger.ts#L1518-L1524

File: src/options/pages/blueprints/BlueprintsToolbar.tsx [hunk 12]
```diff
@@ -97,90 +97,92 @@ const BlueprintsToolbar: React.FunctionComponent<{
...
+            >
+              <FontAwesomeIcon
+                icon={
+                  sortBy[0].id === "updatedAt"
+                    ? sortBy[0].desc
+                      ? faSortAmountDownAlt
+                      : faSortAmountUpAlt
+                    : sortBy[0].desc
+                    ? faSortAmountUpAlt
+                    : faSortAmountDownAlt
```
[the remark this answers is not in the document]
Comment (mnholtz) [reply src/options/pages/blueprints/BlueprintsToolbar.tsx]: I decided to simplify like this: https://github.com/pixiebrix/pixiebrix-extension/pull/4866/commits/a22da42f0dcfc6ef24cc7c71589d4705c477ceb6

:)

**Review by twschiller — APPROVED** [review]

File: src/options/pages/blueprints/BlueprintsToolbar.tsx (hunk 12 above)
[the remark this answers is not in the document]
Comment (fregante) [reply src/options/pages/blueprints/BlueprintsToolbar.tsx]: Even better!

**Comment by fregante** [comment]
One last thing I'd like to see is a white background/box on this view:

Screen Shot


It should be as easy as wrapping it into a Card like I did for the loader in [pixiebrix/pixiebrix-extension@`2d39c76` (#4454)](https://github.com/pixiebrix/pixiebrix-extension/pull/4454/commits/2d39c76c34731edca70f6bc23826891a5351659d)

**Comment by twschiller** [comment]
@mnholtz please merge when this is good to go

**Comment by mnholtz** [comment]
@fregante 

> One last thing I'd like to see is a white background/box on this view

Screen Shot 2023-01-03 at 9 33 40 AM

**Comment by mnholtz** [comment]
@fregante 

Shoot, I just realized that the card component blocks-out the white styling of the kbd styles Gideon designed.  I think I'll revert for now and flag this issue to Corinne.

**Pull request merged**